### PR TITLE
Stop using `addeq!`

### PR DESCRIPTION
**Pull request by SoongNoonien**
This should help with https://github.com/oscar-system/Oscar.jl/pull/4163.

### CODE REVIEW

File: src/GenericCyclotomics.jl [hunk 1]
```diff
@@ -365,10 +365,15 @@ end
 
 mul!(x::GenericCyclo, y::GenericCyclo, z::GenericCyclo) = y*z
 
-add!(x::GenericCyclo, y::GenericCyclo, z::GenericCyclo) = y+z
-
-function addeq!(x::GenericCyclo, y::GenericCyclo)
-	mergewith!(+, x.f, y.f)
+function add!(x::GenericCyclo, y::GenericCyclo, z::GenericCyclo)
+	if x == z
+		y,z=z,y
```
Comment (fingolfin) [inline src/GenericCyclotomics.jl:370]: What if all three are equal?

Comment (SoongNoonien) [reply src/GenericCyclotomics.jl:370]: In this case the swap is unnecessary and useless but not harmful.

Comment (SoongNoonien) [reply src/GenericCyclotomics.jl:370]: But `==` should actually be `===` I think.

Comment (fingolfin) [reply src/GenericCyclotomics.jl:372]: OK, so `mergewith` works correctly if both arguments are the same?

File: src/GenericCyclotomics.jl [hunk 2]
```diff
@@ -367,10 +367,15 @@ end
...
-
-function addeq!(x::GenericCyclo, y::GenericCyclo)
-	mergewith!(+, x.f, y.f)
+function add!(x::GenericCyclo, y::GenericCyclo, z::GenericCyclo)
+	if x === z
+		y,z=z,y
+	else
+		if x !== y
+			x.f = deepcopy(y.f)
+		end
```
Comment (fingolfin) [inline src/GenericCyclotomics.jl:376]:
```suggestion
	elseif x !== y
		x.f = deepcopy(y.f)
```

File: test/runtests.jl [hunk 3]
```diff
@@ -60,6 +60,16 @@ test_Ring_interface(S)
...
+
+	x=deepcopy(b)
+	ab=a+b
+	@test add!(x, a, x) == ab
+	@test x == ab
+
+	add!(a, a, a)
+	@test a == 2*b
+	add!(a, c, b)
+	@test a == c+b
```
Comment (fingolfin) [inline test/runtests.jl:72]: Ideally these tests should also verify that the other arguments are *not* changed.

Comment (fingolfin) [reply test/runtests.jl:72]: On the long rung we should just extend the ring interface tests to cover this, as we want this kind of test for *all* implementations... I.e. `test/Rings-conformance-tests.jl:193` in AA. Not asking you to do this right now, though

Comment (fingolfin) [reply test/runtests.jl:72]: (but see https://github.com/Nemocas/AbstractAlgebra.jl/issues/1814)

File: src/GenericCyclotomics.jl (hunk 1 above)
Comment (SoongNoonien) [reply src/GenericCyclotomics.jl:372]: Yes, I was worried too.
```
julia> a=Dict(["a" => 1, "b" => 2])
Dict{String, Int64} with 2 entries:
  "b" => 2
  "a" => 1

julia> mergewith!(+, a, a)
Dict{String, Int64} with 2 entries:
  "b" => 4
  "a" => 2


```

File: test/runtests.jl (hunk 3 above)
Comment (SoongNoonien) [reply test/runtests.jl:75]: Ok, I've added some extra tests for `add!` at least.

**Review by fingolfin — APPROVED** [review]

**Pull request merged**